From: Michael R Sweet <msweet@msweet.org>
Date: Sat, 28 Nov 2020 08:43:46 -0500
Subject: Force a 5 second sleep to wait for the job control file to be
 written.

---
 test/run-stp-tests.sh | 1 +
 1 file changed, 1 insertion(+)

diff --git a/test/run-stp-tests.sh b/test/run-stp-tests.sh
index 1af6d07..0bf95c5 100755
--- a/test/run-stp-tests.sh
+++ b/test/run-stp-tests.sh
@@ -844,6 +844,7 @@ else
 	echo "PASS"
 	echo "    PASSED" >>$strfile
 
+	sleep 5
 	./waitjobs.sh >>$strfile
 
         echo $ac_n "Verifying that history still exists: $ac_c"
