From: Didier Raboud <odyx@debian.org>
Date: Sat, 14 Nov 2020 18:06:17 +0100
Subject: Set SyncOnClose to @CUPS_SYNC_ON_CLOSE@, not only in comment

---
 conf/cups-files.conf.in | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/conf/cups-files.conf.in b/conf/cups-files.conf.in
index af11fcc..f987c99 100644
--- a/conf/cups-files.conf.in
+++ b/conf/cups-files.conf.in
@@ -7,7 +7,7 @@
 #FatalErrors @CUPS_FATAL_ERRORS@
 
 # Do we call fsync() after writing configuration or status files?
-#SyncOnClose @CUPS_SYNC_ON_CLOSE@
+SyncOnClose @CUPS_SYNC_ON_CLOSE@
 
 # Default user and group for filters/backends/helper programs; this cannot be
 # any user or group that resolves to ID 0 for security reasons...
