From: Thorsten Alteholz <debian@alteholz.de>
Date: Sun, 1 Jun 2025 13:42:36 +0200
Subject: Limit the maximum number of file descriptors to 64k-1

---
 scheduler/main.c | 15 ++++++---------
 1 file changed, 6 insertions(+), 9 deletions(-)

diff --git a/scheduler/main.c b/scheduler/main.c
index ae2409e..c91f477 100644
--- a/scheduler/main.c
+++ b/scheduler/main.c
@@ -519,22 +519,19 @@ main(int  argc,				/* I - Number of command-line args */
 #endif /* HAVE_DBUS_THREADS_INIT */
 
  /*
-  * Set the maximum number of files...
+  * Set the maximum number of files, which for practical reasons can be limited
+  * to the number of TCP port number values (64k-1)...
   */
 
   getrlimit(RLIMIT_NOFILE, &limit);
 
 #if !defined(HAVE_POLL) && !defined(HAVE_EPOLL) && !defined(HAVE_KQUEUE)
-  if (limit.rlim_max > FD_SETSIZE)
+  if ((MaxFDs = limit.rlim_max) > FD_SETSIZE)
     MaxFDs = FD_SETSIZE;
-  else
-#endif /* !HAVE_POLL && !HAVE_EPOLL && !HAVE_KQUEUE */
-#ifdef RLIM_INFINITY
-  if (limit.rlim_max == RLIM_INFINITY)
-    MaxFDs = 16384;
-  else
+#else
+  if ((MaxFDs = limit.rlim_max) > 65535)
+    MaxFDs = 65535;
 #endif /* RLIM_INFINITY */
-    MaxFDs = limit.rlim_max;
 
   limit.rlim_cur = (rlim_t)MaxFDs;
 
