From 3e589a9da248cdfa22f61f20c0957c49f5b9b423 Mon Sep 17 00:00:00 2001
From: Michael R Sweet <michael.r.sweet@gmail.com>
Date: Tue, 20 Feb 2018 09:46:07 -0500
Subject: Fix installation of filters (Issue #5247)

---
 filter/Makefile | 16 ++++++----------
 1 file changed, 6 insertions(+), 10 deletions(-)

diff --git a/filter/Makefile b/filter/Makefile
index 4f67e20e9..b24bc1ee9 100644
--- a/filter/Makefile
+++ b/filter/Makefile
@@ -36,14 +36,14 @@ all:	$(TARGETS)
 # Make library targets...
 #
 
-libs:		$(LIBTARGETS)
+libs:
 
 
 #
 # Make unit tests...
 #
 
-unittests:	$(UNITTARGETS)
+unittests:
 
 
 #
@@ -51,8 +51,7 @@ unittests:	$(UNITTARGETS)
 #
 
 clean:
-	$(RM) $(OBJS) $(TARGETS) $(UNITTARGETS)
-	$(RM) libcupsimage.so libcupsimage.sl libcupsimage.dylib
+	$(RM) $(OBJS) $(TARGETS)
 
 
 #
@@ -83,14 +82,12 @@ install-data:
 
 install-exec:
 	$(INSTALL_DIR) -m 755 $(SERVERBIN)/filter
-	for file in $(FILTERS); do \
+	for file in $(TARGETS); do \
 		$(INSTALL_BIN) $$file $(SERVERBIN)/filter; \
 	done
-	$(RM) $(SERVERBIN)/filter/rastertodymo
-	$(LN) rastertolabel $(SERVERBIN)/filter/rastertodymo
 	if test "x$(SYMROOT)" != "x"; then \
 		$(INSTALL_DIR) $(SYMROOT); \
-		for file in $(FILTERS); do \
+		for file in $(TARGETS); do \
 			cp $$file $(SYMROOT); \
 			dsymutil $(SYMROOT)/$$file; \
 		done \
@@ -116,10 +113,9 @@ install-libs:
 #
 
 uninstall:
-	for file in $(FILTERS); do \
+	for file in $(TARGETS); do \
 		$(RM) $(SERVERBIN)/filter/$$file; \
 	done
-	$(RM) $(SERVERBIN)/filter/rastertodymo
 	-$(RMDIR) $(SERVERBIN)/filter
 	-$(RMDIR) $(SERVERBIN)
 
