--- a/conf/mime.types
+++ b/conf/mime.types
@@ -92,7 +92,7 @@
 				 char(3,0xe4) char(3,0xe5) char(3,0xe6) char(3,0xe7)\
 				 char(3,0xe8) char(3,0xe9) char(3,0xea) char(3,0xeb)\
 				 char(3,0xec) char(3,0xed) char(3,0xee) char(3,0xef))
-image/pwg-raster		string(0,"RaS2") + string(4,PwgRaster<00>) priority(100)
+image/pwg-raster		string(0,"RaS2") + string(4,PwgRaster<00>) priority(150)
 image/tiff			tiff tif string(0,MM<002A>) string(0,II<2A00>)
 image/x-photocd			pcd string(2048,PCD_IPI)
 image/x-portable-anymap		pnm
