--- a/backend/usb-libusb.c
+++ b/backend/usb-libusb.c
@@ -361,25 +361,23 @@
 
             if (!open_device(&printer, data != NULL))
 	    {
-	      if (!get_device_id(&printer, device_id, sizeof(device_id)))
-	      {
-                make_device_uri(&printer, device_id, device_uri,
-		                sizeof(device_uri));
+	      get_device_id(&printer, device_id, sizeof(device_id));
+	      make_device_uri(&printer, device_id, device_uri,
+			      sizeof(device_uri));
 
-	        if ((*cb)(&printer, device_uri, device_id, data))
-		{
-		  printer.read_endp  = printer.device->config[printer.conf].
+	      if ((*cb)(&printer, device_uri, device_id, data))
+	      {
+		printer.read_endp  = printer.device->config[printer.conf].
 				           interface[printer.iface].
 					   altsetting[printer.altset].
 					   endpoint[printer.read_endp].
 					   bEndpointAddress;
-		  printer.write_endp = printer.device->config[printer.conf].
+		printer.write_endp = printer.device->config[printer.conf].
 					   interface[printer.iface].
 					   altsetting[printer.altset].
 					   endpoint[printer.write_endp].
 					   bEndpointAddress;
-		  return (&printer);
-		}
+		return (&printer);
               }
 
               close_device(&printer);
@@ -481,7 +479,8 @@
   * Get the device URI and make/model strings...
   */
 
-  backendGetMakeModel(device_id, make_model, sizeof(make_model));
+  if (backendGetMakeModel(device_id, make_model, sizeof(make_model)))
+    strlcpy(make_model, "Unknown", sizeof(make_model));
 
  /*
   * Report the printer...
@@ -514,7 +513,7 @@
   cups_option_t	*values;		/* 1284 parameters */
   const char	*mfg,			/* Manufacturer */
 		*mdl,			/* Model */
-		*des,			/* Description */
+		*des = NULL,		/* Description */
 		*sern;			/* Serial number */
   size_t	mfglen;			/* Length of manufacturer string */
   char		tempmfg[256],		/* Temporary manufacturer string */
@@ -596,6 +595,19 @@
     mfg = tempmfg;
   }
 
+  if (!mdl)
+  {
+   /*
+    * No model?  Use description...
+    */
+    if (des)
+      mdl = des; /* We remove the manufacturer name below */
+    else if (!strncasecmp(mfg, "Unknown", 7))
+      mdl = "Printer";
+    else
+      mdl = "Unknown Model";
+  }
+
   mfglen = strlen(mfg);
 
   if (!strncasecmp(mdl, mfg, mfglen) && _cups_isspace(mdl[mfglen]))
